File: go-code/refactoring_api/separate_query_from_modifier/after/alarm.go
package before

// STEP 1: Create a new function to implement the query aspect
func findMiscreant(people []string) string {
	for _, p := range people {
		if p == "Don" || p == "John" {
			return p
		}
	}
	return ""
}

// STEP2: Refactor the alert function to convert it into a command function
// STEP2.1: Remove return type
// STEP 2.2: Remove query logic and escalate it to the query method
func alertForMiscreant(people []string) {
	if findMiscreant(people) != "" {
		setOffAlarms()
	}
}

func setOffAlarms() {
	// Implement setOffAlarms logic if needed
}

func main() {
	people := []string{"Alice", "Bob", "Don", "Eve", "John"}
	//STEP 2.3: First call the query. As you can see there is duplication here. findMiscreant()
	// function is called both by the client and in the API. That is ok
	alertedPerson := findMiscreant(people)
	alertForMiscreant(people)
	if alertedPerson != "" {
		println("Alert for:", alertedPerson)
	} else {
		println("No alert triggered.")
	}
}

File: java-code/src/main/java/refactoring/refactoring_api/separate_query_from_modifier/after/Alarm.java
package refactoring.refactoring_api.separate_query_from_modifier.after;

import java.util.List;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class Alarm {

    public static void main(String[] args) {
      List<String> people = List.of("Alice", "Bob", "Don", "Eve", "John");
      var alarm = new Alarm();
      //STEP 2.2: First call the query method and stop using return value of the command method
      var alertedPerson = alarm.findAlarmedPerson(people);
      alarm.alertForMiscreant(people);

      if (!alertedPerson.isEmpty()) {
        log.info("Alert for: {}",  alertedPerson);
      } else {
        log.info("No alert triggered.");
      }
    }

  //STEP 1: Copy the method and refactor it to a query method
    public String findAlarmedPerson(List<String> people) {
    for (String p : people) {
      if ("Don".equals(p) || "John".equals(p)) {
        return p;
      }
    }
    return "";
  }

  //STEP 2: Convert the method into a command method
  //STEP 2.1: Remove return type
  public void alertForMiscreant(List<String> people) {
      //STEP 2.3: Remove query logic from command and call query method directly
    //query method (findAlarmedPerson)  is called both from the client and the command method.
    // This may look like a duplication, but it is ok to have it like that.
      if(!"".equals(findAlarmedPerson(people))) {
        setOffAlarms();
      }
    }

    public void setOffAlarms() {
      // Implement setOffAlarms logic if needed
    }
  
}
